src/conv/heuristics/ai_heuristics.cpp

```diff
@@ -0,0 +1,455 @@
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/conv/heuristics/ai_heuristics.hpp>
+#if MIOPEN_ENABLE_AI_IMMED_MODE_FALLBACK || MIOPEN_ENABLE_AI_KERNEL_TUNING
+#include <fdeep/fdeep.hpp>
+#include <filesystem>
+
+namespace miopen {
+namespace ai {
+namespace common {
+nlohmann::json LoadJSON(const std::string& path)
+{
+    if(!std::filesystem::exists(path))
+        MIOPEN_THROW(miopenStatusInternalError, "Unable to load file: " + path);
+    return nlohmann::json::parse(std::ifstream(path));
+}
+template <typename U, typename V>
+std::unordered_map<V, U> ReverseMap(const std::unordered_map<U, V>& map)
+{
+    std::unordered_map<V, U> reversed_map = {};
+    for(const auto& it : map)
+        reversed_map.emplace(make_pair(it.second, it.first));
+    return reversed_map;
+}
+template <typename U, typename V>
+std::vector<V> LookupValues(const std::vector<U>& keys, const std::unordered_map<U, V>& map)
+{
+    std::vector<V> values = {};
+    values.reserve(keys.size());
+    std::transform(keys.begin(), keys.end(), std::back_inserter(values), [&](const U& key) {
+        return map.at(key);
+    });
+    return values;
+}
+} // namespace common
+
+#if MIOPEN_ENABLE_AI_IMMED_MODE_FALLBACK
+namespace immed_mode {
+Metadata::Metadata(const std::string& arch)
+    : json(common::LoadJSON(GetSystemDbPath() + "/" + arch + "_metadata.tn.model")),
+      direction_encodings(json["encodings"]["Direction"]),
+      precision_encodings(json["encodings"]["Precision"]),
+      layout_encodings(json["encodings"]["Layout"]),
+      features(json["conv_params_used_as_features"]),
+      num_inputs(json["num_inputs"]),
+      num_outputs(json["num_outputs"]),
+      num_solvers(json["num_solvers"]),
+      solver_map(common::ReverseMap<std::string, size_t>(json["encodings"]["solver"])),
+      features_mean(common::LookupValues<std::string, float>(
+          features, json["stats"]["overall"]["features"]["mean"])),
+      features_std(common::LookupValues<std::string, float>(
+          features, json["stats"]["overall"]["features"]["std"]))
+{
+}
+
+size_t Metadata::EncodeDirection(miopen::conv::Direction dir) const
+{
+    if(dir == conv::Direction::BackwardWeights)
+        return direction_encodings.at("W");
+    else if(dir == conv::Direction::BackwardData)
+        return direction_encodings.at("B");
+    else
+        return direction_encodings.at("F");
+}
+
+size_t Metadata::EncodePrecision(miopenDataType_t data_type) const
+{
+    if(data_type == miopenBFloat16)
+        return precision_encodings.at("BF16");
+    else if(data_type == miopenHalf)
+        return precision_encodings.at("FP16");
+    else if(data_type == miopenFloat)
+        return precision_encodings.at("FP32");
```

Comment:
```suggestion
    switch(data_type)
    {
    case miopenBFloat16:
        return precision_encodings.at("BF16");
    case miopenHalf:
        return precision_encodings.at("FP16");
    case miopenFloat:
        return precision_encodings.at("FP32");
    }
```

Reply:
If I make this change I get this compiler warning:

4 enumeration values not explicitly handled in switch: 'miopenInt32', 'miopenInt8', 'miopenInt8x4'...

Reply:
Use `default` and throw the exception in default case.